appi.adoc

```diff
@@ -0,0 +1,397 @@
+
+This appendix contains recommended practices for using netCDF-4 groups in CF compliant data, as well as annotated examples of existing use cases for representing data using groups.
+
+=== Recommended practices
+
+. The use of attributes attached to a group in order to store metadata normally attached directly to variables is discouraged.
+This includes, for example, replacing per-variable attributes like `_FillValue`, `scale_factor`, `valid_min`, with group-level equivalents.
+Although group attributes might be more concise, it is likely to create problems with downstream software and to reduce interoperability.
+
+. Each group's attributes, dimensions and variables should be self-contained in combination with the group metadata and any inherited properties (e.g., dimension sizes, coordinates).
+This ensures that if a group is renamed or extracted (with any inherited properties) into a new file, its information content is preserved.
+For example, storage of ensembles as sibling groups (as described below) is often clearer when the realization number is encoded in the group name.
+In this case, the `Realization` group attribute retains the realization number even if the group is renamed.
+Other commonly enumerated group names, such as station identifiers, buoy numbers, or channel wavelengths are analogous.
+It is fine to enumerate or itemize names so long as the number or name is redundantly stored as a group attribute.
+
+. Renaming or moving a group or self-contained branch of groups to a new location should not affect the interpretation of data.
+Relative and absolute paths (containing "/") of coordinates, dimensions, and attributes are fragile in that they may break if groups are moved or renamed.
+To prevent this, it is preferred that paths not be present in attribute values.
+Instead, named objects resolve to the most proximal object (i.e., dimension or variable) of that name that has the referring attribute within its scope.
+This makes files which use groups amenable to sub-setting and aggregation.
+For example, the CF `coordinates` attribute identifies a variable's coordinates in a whitespace-separated list such as `"lat lon"`.
+When the coordinates are outside the group that contains the `coordinates` attribute, it is tempting to store the coordinate locations as full, unambiguous paths such as `"/g1/lat /g1/lon"`, or as relative paths such as `"g1/lat g1/lon"`.
+However, paths that contain slashes must be explicitly altered when the variable is subset into a new file with a different group hierarchy, or when the hierarchy is flattened.
+CF attributes affected by this practice include `ancillary_variables`, `bounds`, `cell_measures`, `climatology`,  `coordinates`, `formula_terms`, and `grid_mapping`.
+The simpler `"lat lon"` specification works without alteration in all situations for out-of-group locations so long as it is understood to mean the nearest identifiers that have the referring variable in their scope.
```

Comment:
Just a question: is `coordinates="/g1/lat /g1/lon"` _ever_ a good idea, in practice? You're argument that the search rules should be instead in preference is good, and I can't think of any exceptions, off hand. Perhaps there are legacy datasets that need catering for?

Reply:
@davidhassell Your comment is slightly garbled. Many NASA legacy datasets currently use the `"/g1/lat /g1/lon"` paradigm. The lateral search algorithm would allow them to maintain the same group structure while eliminating the absolute paths.

Reply:
OK - no problem, apologies for not making myself clearer. I consider the question answered, and the text is fine.